File: python/day3/roman.py
#!/usr/bin/env python

# Roman numbers (with some meta programming)

class ClassAttributeInterceptor(type):

    def __getattr__(cls, name):
        return cls.classfeature_missing(name)
        
    def classfeature_missing(cls, name):
        pass
        

class Roman(object):

    __metaclass__ = ClassAttributeInterceptor

    @classmethod
    def classfeature_missing(cls, name):
        return Roman.to_int(name)
                    
    @classmethod
    def to_int(cls, roman):
        roman = roman.replace('IV', 'IIII')
        roman = roman.replace('IX', 'VIII')
        roman = roman.replace('XL', 'XXXX')
        roman = roman.replace('XC', 'LXXXX')
        roman = roman.replace('CD', 'CCCC')
        roman = roman.replace('CM', 'DCCCC')
    
        result = (roman.count('I')
            + roman.count('V') * 5
            + roman.count('X') * 10
            + roman.count('L') * 50
            + roman.count('C') * 100
            + roman.count('D') * 500
            + roman.count('M') * 1000)
        return result
    

# Regular class method calls
print "%s" % Roman.to_int('X')
print "%s" % Roman.to_int('XC')
print "%s" % Roman.to_int('MCMLXXXIV')

# via ClassAttributeInterceptor
print "%s" % Roman.MCMLXXXIV

File: ruby/day3/roman.rb
#!/usr/bin/env ruby

# Roman numbers (with some meta programming)

class Roman

    # Override class method method_missing.
    # The method name will be used as the input parameter for the number conversion
    def self.method_missing(name, *args)
        return to_i(name.to_s)
    end
    
    # Convert a string with some roman number into an integer
    def self.to_i(roman)
        roman.gsub!('IV', 'IIII')
        roman.gsub!('IX', 'VIII')
        roman.gsub!('XL', 'XXXX')
        roman.gsub!('XC', 'LXXXX')
        roman.gsub!('CD', 'CCCC')
        roman.gsub!('CM', 'DCCCC')
    
        puts "#{roman}"
        
        # Multiline statement, be sure to end each line with an operator or \
        result = (roman.count("I") +
            roman.count("V") * 5 +
            roman.count("X") * 10 +
            roman.count("L") * 50 +
            roman.count("C") * 100 +
            roman.count("D") * 500 +
            roman.count("M") * 1000)
        return result
    end  
end  

# Regular class method calls
puts Roman.to_i('XI')
puts Roman.to_i('XC')
puts Roman.to_i('MCMLXXXIV')

# via overriding method_missing
puts Roman.MCMLXXXIV

